File: python/structures/src/searches/graph/bfs.py
def traverse(root):
    output = ''

    queue = [root]
    visited = []

    while len(queue) > 0:
        # pop the next item in the queue
        curr = queue.pop(0)

        # add all unvisited neighbors to the queue
        for neighbor in curr.neighbors:
            if neighbor not in visited and neighbor not in queue:
                queue.append(neighbor)

        output += str(curr.value) + ' '

        visited.append(curr)

    return output.strip()

File: ruby/src/searches/graph/bfs.rb
require_relative '../../graph/directed'

module BreadthFirst
    def BreadthFirst.traverse(root)
        out = ''
        queue = [root]
        visited = []

        while !queue.empty?
            curr = queue.shift

            out += curr.value.to_s + ' '

            for neighbor in curr.neighbors
                if !queue.include?(neighbor) && !visited.include?(neighbor)
                    queue << neighbor
                end
            end

            visited << curr
        end

        return out.strip
    end
end
